internalservice_agentmanager.go

```diff
@@ -0,0 +1,331 @@
+	watchersCancel  func()
+	needsBucketChan chan struct{}
+
+	topologyCfgWatcher *ConfigWatcherHttp
+}
+
+type bucketDescriptor struct {
+	Name string
+	UUID string
+}
+
+func CreateInternalServiceAgentManager(ctx context.Context, opts InternalServiceAgentManagerOptions) (*InternalServiceAgentManager, error) {
+	srcHTTPAddrs := makeSrcHTTPAddrs(opts.SeedConfig.HTTPAddrs, opts.TLSConfig)
+	httpTransport := makeHTTPTransport(opts.TLSConfig)
+	httpUserAgent := "gocbcorex/0.0.1-dev"
+
+	logger := loggerOrNop(opts.Logger)
+	logger = logger.Named("agent manager")
+
+	bootstrapper, err := NewConfigBootstrapHttp(ConfigBoostrapHttpOptions{
+		Logger:           logger.Named("http-bootstrap"),
+		HttpRoundTripper: httpTransport,
+		Endpoints:        srcHTTPAddrs,
+		UserAgent:        httpUserAgent,
+		Authenticator:    opts.Authenticator,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	bootstrapConfig, networkType, err := bootstrapper.Bootstrap(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Debug("internal service agent manager bootstrapped",
+		zap.Any("bootstrapConfig", bootstrapConfig),
+		zap.String("networkType", networkType))
+
+	m := &InternalServiceAgentManager{
+		opts:         opts,
+		logger:       logger,
+		userAgent:    httpUserAgent,
+		networkType:  networkType,
+		bucketAgents: make(map[string]*Agent),
+		state: &internalServiceAgentManagerState{
+			tlsConfig:     opts.TLSConfig,
+			authenticator: opts.Authenticator,
+			httpTransport: httpTransport,
+			latestConfig:  bootstrapConfig,
+		},
+	}
+
+	bucketsWatcher, err := NewStreamWatcherHttp[[]bucketDescriptor](&StreamWatcherHttpConfig{
+		HttpRoundTripper: httpTransport,
+		Endpoints:        srcHTTPAddrs,
+		UserAgent:        httpUserAgent,
+		Authenticator:    opts.Authenticator,
+	}, &StreamWatcherHttpOptions{
+		Logger: logger.Named("buckets-watcher"),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	m.bucketsWatcher = bucketsWatcher
+
+	topoWatcher, err := NewConfigWatcherHttp(
+		&ConfigWatcherHttpConfig{
+			HttpRoundTripper: httpTransport,
+			Endpoints:        srcHTTPAddrs,
+			UserAgent:        httpUserAgent,
+			Authenticator:    opts.Authenticator,
+		},
+		&ConfigWatcherHttpOptions{
+			Logger: logger.Named("topology-watcher"),
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	m.topologyCfgWatcher = topoWatcher
+
+	clusterAgent, err := m.makeAgentLocked(ctx, "")
+	if err != nil {
+		return nil, err
+	}
+
+	m.clusterAgent = clusterAgent
+
+	m.startWatchers()
+
+	return m, nil
+}
+
+func (m *InternalServiceAgentManager) startWatchers() {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	stoppedTopoSig := make(chan struct{})
+	stoppedBucketSig := make(chan struct{})
+	m.watchersCancel = func() {
+		// Cancel the context and wait for the configCh to be closed.
+		cancel()
+		<-stoppedBucketSig
+		<-stoppedTopoSig
+	}
+
+	bucketCh := m.bucketsWatcher.Watch(ctx, streamWatcherHttp_streamBuckets)
+
+	go func() {
+		for bucket := range bucketCh {
```

Comment:
`bucket` here is actually a list of buckets I think?  This should probably be called `buckets`.